### PR TITLE
Add TSDF RangeDataInserter

### PR DESCRIPTION
**Pull request by kdaun**
Adds TSDF RangeDataInserter, tests and integrates the configuration files.

### CODE REVIEW

File: cartographer/mapping/2d/tsdf_range_data_inserter_2d.cc [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    }
+    float weight_factor_range = 1.f;
+    if (options_.update_weight_range_exponent() != 0) {
+      weight_factor_range =
+          ComputeWeight(range, options_.update_weight_range_exponent());
+    }
+
+    // Update Cells.
+    for (const Eigen::Array2i& cell_index : ray_mask) {
+      Eigen::Vector2f cell_center = tsdf->limits().GetCellCenter(cell_index);
```
Comment (cschuet) [inline cartographer/mapping/2d/tsdf_range_data_inserter_2d.cc]: check update marker for early return

File: cartographer/mapping/2d/tsdf_range_data_inserter_2d.cc [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+  float updated_weight = tsd_and_weight.second + update_weight;
+  float updated_sdf = (tsd_and_weight.first * tsd_and_weight.second +
+                       update_sdf * update_weight) /
+                      updated_weight;
+  updated_weight =
+      std::min(updated_weight, static_cast<float>(options_.maximum_weight()));
+  tsdf->SetCell(cell, updated_sdf, updated_weight);
+}
+
+float TSDFRangeDataInserter2D::ComputeWeight(float ray_length,
```
Comment (cschuet) [inline cartographer/mapping/2d/tsdf_range_data_inserter_2d.cc]: rename

File: cartographer/mapping/2d/tsdf_range_data_inserter_2d.cc [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+  return options;
+}
+
+TSDFRangeDataInserter2D::TSDFRangeDataInserter2D(
+    const proto::TSDFRangeDataInserterOptions2D& options)
+    : options_(options) {}
+
+// Casts a ray from origin towards hit for each hit in range data.
+// If 'options.update_free_space' is 'true', all cells along the ray
+// until 'truncation_distance' behind hit ar updated. Otherwise, only the cells
```
Comment (cschuet) [inline cartographer/mapping/2d/tsdf_range_data_inserter_2d.cc]: s/ar/are/g

File: cartographer/mapping/2d/tsdf_range_data_inserter_2d_test.cc [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+    // Cells intersect with ray.
+    float x = -0.5f;
+    Eigen::Array2i cell_index =
+        tsdf_.limits().GetCellIndex(Eigen::Vector2f(x, y));
+    float expected_tsdf =
+        std::max(std::min(3.5f - y, truncation_distance), -truncation_distance);
+    float expected_weight = 1.f;
+    EXPECT_THAT(MockCellProperties(cell_index, tsdf_),
+                EqualCellProperties(true, expected_tsdf, expected_weight));
+    // Cells don't intersect with ray..
```
Comment (cschuet) [inline cartographer/mapping/2d/tsdf_range_data_inserter_2d_test.cc]: remove additional dot here and everywhere

File: cartographer/mapping/proto/2d/tsdf_range_data_inserter_options_2d.proto [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+import "cartographer/mapping/proto/2d/normal_estimation_options_2d.proto";
+
+message TSDFRangeDataInserterOptions2D {
+  // Distance to the surface within the signed distance function is evaluated
+  double truncation_distance = 1;
+  // Maximum weight that can be stored in a cell
+  double maximum_weight = 2;
+
+  bool update_free_space = 3;
```
Comment (cschuet) [inline cartographer/mapping/proto/2d/tsdf_range_data_inserter_options_2d.proto]: Add comment

**Comment by kdaun** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user kdaun. Merge queue now has a length of 3.

**Comment by kdaun** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user kdaun. Merge queue now has a length of 5.

**Comment by cschuet** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user cschuet. Merge queue now has a length of 6.

**Comment by wally-the-cartographer** [comment]
Error: exit status 1

Log:
~~~
=> handle work item: PR 1236 on googlecartographer/cartographer.
 => Switching to branch master.
 $ git reset --hard HEAD [data/cartographer]
HEAD is now at 7dfe4042 Add TSDF RangeDataInserter (#1236)
 $ git clean -fd [data/cartographer]
 $ git checkout master [data/cartographer]
Already on 'master'
Your branch is up to date with 'origin/master'.
 $ git pull [data/cartographer]
From https://github.com/googlecartographer/cartographer
   7dfe4042..eff103f6  master     -> origin/master
Updating 7dfe4042..eff103f6
Fast-forward
 cartographer/cloud/internal/handlers/add_trajectory_handler.cc      | 3 +++
 cartographer/cloud/internal/handlers/add_trajectory_handler_test.cc | 2 ++
 2 files changed, 5 insertions(+)
Current branch master is up to date.
 => Checking out kdaun/integrate_range_data_inserter as local branch pr_1236.
 $ git remote add kdaun git@github.com:kdaun/cartographer.git [data/cartographer]
fatal: remote kdaun already exists.
 $ git fetch kdaun [data/cartographer]
 $ git branch --track pr_1236 kdaun/integrate_range_data_inserter [data/cartographer]
Branch 'pr_1236' set up to track remote branch 'integrate_range_data_inserter' from 'kdaun'.
 => Switching to branch pr_1236.
 $ git reset --hard HEAD [data/cartographer]
HEAD is now at eff103f6 Ignore initial poses on gRPC server side. (#1234)
 $ git clean -fd [data/cartographer]
 $ git checkout pr_1236 [data/cartographer]
Switched to branch 'pr_1236'
Your branch is up to date with 'kdaun/integrate_range_data_inserter'.
 $ git pull --force [data/cartographer]
Already up to date.
Current branch pr_1236 is up to date.
 => merge master.
 $ git merge master [data/cartographer]
Merge made by the 'recursive' strategy.
 cartographer/cloud/internal/handlers/add_trajectory_handler.cc      | 3 +++
 cartographer/cloud/internal/handlers/add_trajectory_handler_test.cc | 2 ++
 2 files changed, 5 insertions(+)
 => Running clang-format on all *.cc and *.h files
 $ git commit -am Ran clang-format. [data/cartographer]
On branch pr_1236
Your branch is ahead of 'kdaun/integrate_range_data_inserter' by 3 commits.
  (use "git push" to publish your local commits)

nothing to commit, working tree clean
 $ git push --force kdaun HEAD:integrate_range_data_inserter [data/cartographer]
To github.com:kdaun/cartographer.git
 ! [remote rejected]   HEAD -> integrate_range_data_inserter (permission denied)
error: failed to push some refs to 'git@github.com:kdaun/cartographer.git'
 => Abandoning branch pr_1236.
 => Switching to branch master.
 $ git reset --hard HEAD [data/cartographer]
HEAD is now at 997268be Merge branch 'master' into pr_1236
 $ git clean -fd [data/cartographer]
 $ git checkout master [data/cartographer]
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
 $ git branch -D pr_1236 [data/cartographer]
Deleted branch pr_1236 (was 997268be).

~~~

**Comment by wally-the-cartographer** [comment]
Error: exit status 1

Log:
~~~
=> handle work item: PR 1236 on googlecartographer/cartographer.
 => Switching to branch master.
 $ git reset --hard HEAD [data/cartographer]
HEAD is now at eff103f6 Ignore initial poses on gRPC server side. (#1234)
 $ git clean -fd [data/cartographer]
 $ git checkout master [data/cartographer]
Already on 'master'
Your branch is up to date with 'origin/master'.
 $ git pull [data/cartographer]
Already up to date.
Current branch master is up to date.
 => Checking out kdaun/integrate_range_data_inserter as local branch pr_1236.
 $ git remote add kdaun git@github.com:kdaun/cartographer.git [data/cartographer]
fatal: remote kdaun already exists.
 $ git fetch kdaun [data/cartographer]
 $ git branch --track pr_1236 kdaun/integrate_range_data_inserter [data/cartographer]
Branch 'pr_1236' set up to track remote branch 'integrate_range_data_inserter' from 'kdaun'.
 => Switching to branch pr_1236.
 $ git reset --hard HEAD [data/cartographer]
HEAD is now at eff103f6 Ignore initial poses on gRPC server side. (#1234)
 $ git clean -fd [data/cartographer]
 $ git checkout pr_1236 [data/cartographer]
Switched to branch 'pr_1236'
Your branch is up to date with 'kdaun/integrate_range_data_inserter'.
 $ git pull --force [data/cartographer]
Already up to date.
Current branch pr_1236 is up to date.
 => merge master.
 $ git merge master [data/cartographer]
Merge made by the 'recursive' strategy.
 cartographer/cloud/internal/handlers/add_trajectory_handler.cc      | 3 +++
 cartographer/cloud/internal/handlers/add_trajectory_handler_test.cc | 2 ++
 2 files changed, 5 insertions(+)
 => Running clang-format on all *.cc and *.h files
 $ git commit -am Ran clang-format. [data/cartographer]
On branch pr_1236
Your branch is ahead of 'kdaun/integrate_range_data_inserter' by 3 commits.
  (use "git push" to publish your local commits)

nothing to commit, working tree clean
 $ git push --force kdaun HEAD:integrate_range_data_inserter [data/cartographer]
To github.com:kdaun/cartographer.git
 ! [remote rejected]   HEAD -> integrate_range_data_inserter (permission denied)
error: failed to push some refs to 'git@github.com:kdaun/cartographer.git'
 => Abandoning branch pr_1236.
 => Switching to branch master.
 $ git reset --hard HEAD [data/cartographer]
HEAD is now at fb2e85d9 Merge branch 'master' into pr_1236
 $ git clean -fd [data/cartographer]
 $ git checkout master [data/cartographer]
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
 $ git branch -D pr_1236 [data/cartographer]
Deleted branch pr_1236 (was fb2e85d9).

~~~

**Pull request merged**